File: 2025/day09.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <cstdint>
#include <cstring>
#include <cmath>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
    "7,1\n"
    "11,1\n"
    "11,7\n"
    "9,7\n"
    "9,5\n"
    "2,5\n"
    "2,3\n"
    "7,3"
);

#define DAY "day09"

bool DEBUG = false;
bool TEST = false;

typedef Point<int> point_t;

int64_t part1 ( const vector<point_t> & data )
{
    int64_t best = 0;
    for( int i = 0; i < data.size(); i++ )
        for( int j = i+1; j < data.size(); j++ )
        {
            int64_t area =
                int64_t(abs(data[i].x-data[j].x)+1) *
                int64_t(abs(data[i].y-data[j].y)+1);
            best = max(area, best);
        }
    return best;
}

int64_t part2 ( const vector<point_t> & data )
{
    int64_t best = 0;
    for( int i = 0; i < data.size(); i++ )
        for( int j = i+1; j < data.size(); j++ )
        {
            int64_t area =
                int64_t(abs(data[i].x-data[j].x)+1) *
                int64_t(abs(data[i].y-data[j].y)+1);

            if( area < best )
                continue;

            int x1 = min( data[i].x, data[j].x );
            int x2 = max( data[i].x, data[j].x );
            int y1 = min( data[i].y, data[j].y );
            int y2 = max( data[i].y, data[j].y );

            bool maybe = true;
            for( int k = 0; k < data.size() - 1; k++ )
            {
                point_t p1 = data[k];
                point_t p2 = data[k+1];
                if( p1.x == p2.x )
                {
                    int py0 = min( p1.y, p2.y );
                    int py1 = max( p1.y, p2.y );
                    if( x1 < p1.x && p1.x < x2 && py0 <= y2 && py1 >= y1 )
                    {
                        maybe = false;
                        break;
                    }
                } else {
                    // Horizontal.
                    int px0 = min( p1.x, p2.x );
                    int px1 = max( p1.x, p2.x );
                    if( y1 < p1.y && p1.y < y2 && px0 <= x2 && px1 >= x1 )
                    {
                        maybe = false;
                        break;
                    }
                }
            }
            if( maybe )
            {
                if( DEBUG )
                    cout << area << "\t" << data[i].x << "," << data[i].y << " " << data[j].x << "," << data[j].y << "\n";
                best = area;
            }
    }
    return best;
}

int main( int argc, char ** argv )
{
    string name = *argv;
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents(DAY".txt");

    // Construct points.

    vector<point_t> points;
    for( auto & line : split(input) )
    {
        auto p = split(line,",");
        points.emplace_back(point_t{
            stoi(p[0]),
            stoi(p[1])
        });
    }

    cout << "Part 1: " << part1(points) << "\n";
    cout << "Part 2: " << part2(points) << "\n";
}

File: 2025/day09.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
7,1
11,1
11,7
9,7
9,5
2,5
2,3
7,3`[1:]

//go:embed day09.txt
var live string

type Point = tools.Point

func part1(data []Point) int {
    best := 0
	for i, a := range data {
		for _, b := range data[i+1:] {
			area := (tools.AbsInt(b.X-a.X)+1) * (tools.AbsInt(b.Y-a.Y)+1)
			if area > best {
				best = area
			}
		}
	}
    return best
}

func part2(data []Point) int {
    best := 0
	for i, a := range data {
		for _, b := range data[i+1:] {
			area := (tools.AbsInt(b.X-a.X)+1) * (tools.AbsInt(b.Y-a.Y)+1)
			if area < best {
				continue
			}
			x1 := a.X
			x2 := b.X
			if x1 > x2 {
				x1, x2 = x2, x1
			}
			y1 := a.Y
			y2 := b.Y
			if y1 > y2 {
				y1, y2 = y2, y1
			}
			
			maybe := true
			for i := 0; i < len(data)-1; i++ {
				p1 := data[i];
				p2 := data[i+1];
				if p1.X == p2.X {
					// Vertical.
					py0 := p1.Y
					py1 := p2.Y
					if py0 > py1 {
						py0, py1 = py1, py0
					}
					if x1 <  p1.X && p1.X < x2 && py0 <= y2 && py1 >= y1 {
						maybe = false
						break
					}
				} else {
					// Horizontal.
					px0 := p1.X;
					px1 := p2.X;
					if px0 > px1 {
						px0, px1 = px1, px0
					}
					if y1 < p1.Y && p1.Y < y2 && px0 <= x2 && px1 >= x1 {
						maybe = false
						break
					}
				}
			}
			if maybe {
				if DEBUG {
					fmt.Println( area, a, b )
				}
				best = area
			}
		}
	}
	return best
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := []Point{}
	for _, row := range strings.Split(input, "\n") {
		p := strings.Split(row, ",")
		data = append(data, Point{
			tools.StrToInt(p[0]),
			tools.StrToInt(p[1])})
	}

	fmt.Println("Part 1:", part1(data))
	fmt.Println("Part 2:", part2(data))
}
